README
NAME
    WebService::ForecastIO - Perl client for api.forecast.io

VERSION
    version 0.01

SYNOPSIS
        use 5.014;

        my $forecast = WebService::ForecastIO->new(
                api_key => 'secret',
                units => 'si', # metric units
                exclude => 'hourly,minutely,currently,flags', # get 7 day forecast only
        );

        #                             latitude          longitude
        my $json = $forecast->request(29.7492738082192, -95.4709680410959);

        # Human readable summary of daily data
        say $json->{daily}->{summary};

OVERVIEW
    This is a Perl client for forecast.io API
    <https://developer.forecast.io>. Forecast.io applies "big data" analysis
    techniques to publicly available weather data including radar image
    analysis. One of the things it attempts to predict is *when* certain
    weather events like rain will start, and the duration of those events.

    This library requires an API key which can be obtained for free from the
    developer web site <https://developer.forecast.io>. The first 1,000
    calls per day are allowed without charge. (More calls can be made if
    payment arrangements are made.)

    See the API docs <https://developer.forecast.io/docs/v2> for full
    details about what data is provided and what granularity data sets are
    offered.

    NOTE: Errors are fatal. Please use something like Try::Tiny if you want
    to handle errors some other way.

ATTRIBUTES
  units
    This attribute overrides the default units of "us". Available units are:

    *   "si"

        Metric units (windspeed in meters/second)

    *   "ca"

        Metric units, except windspeed in km/h

    *   "uk"

        Metric units, except windspeed in mi/h

    *   "auto"

        Unit selection based on request IP geolocation

  exclude
    A comma delimited list of data sets to exclude. Choices are:

    "minutely", "currently", "hourly", "daily", "alerts", "flags"

    See the API docs for information about what these data blocks represent.

    Spaces should be omitted between exclusions.

  api_key
    Holds the API key for the forecast.io service.

METHODS
  request
    This method takes two or three parameters.

    The first two parameters are "latitude" and "longitude" expressed as
    floats. The optional third parameter is a "time" value. The time value
    can be expressed as either epoch seconds or as an iso8601 format time.

    Data sets up to 60 years old are available. (Such times should be
    expressed as iso8601 format times since epoch seconds start at midnight,
    January 1, 1970.)

  to_timepiece
    A convenience method that takes epoch seconds and returns a new
    Time::Piece object. You can stringify this object to iso8601 format by
    calling the "datetime()" method on it.

  parse_datetime
    A convenience method that takes an iso8601 formatted string and returns
    a new Time::Piece object. This object can output epoch seconds by
    calling the "epoch()" method on it.

SEE ALSO
    *   WebService::ForecastIO::Request

    *   API docs <https://developer.forecast.io/docs/v2>

AUTHOR
    Mark Allen <mrallen1@yahoo.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2013 by Mark Allen.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Change history for WebService-ForecastIO

0.01        2013-05-08
            Initial release


Makefile.PL

use strict;
use warnings;

use 5.014;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Perl client for api.forecast.io",
  "AUTHOR" => "Mark Allen <mrallen1\@yahoo.com>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "WebService-ForecastIO",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "WebService::ForecastIO",
  "PREREQ_PM" => {
    "HTTP::Tiny" => 0,
    "JSON" => 0,
    "Moo" => 0,
    "Moo::Role" => 0,
    "Time::Piece" => 0
  },
  "TEST_REQUIRES" => {
    "Socket" => 0,
    "Test::More" => 0,
    "strict" => 0
  },
  "VERSION" => "0.01",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  my $tr = delete $WriteMakefileArgs{TEST_REQUIRES};
  my $br = $WriteMakefileArgs{BUILD_REQUIRES};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/WebService/ForecastIO/Request.pm
use 5.014;

package WebService::ForecastIO::Request;
{
  $WebService::ForecastIO::Request::VERSION = '0.01';
}

use Moo::Role;
use HTTP::Tiny;
use JSON;

# ABSTRACT: Request role for WebService::ForecaseIO


has 'base_url' => (
    is => 'ro',
    default => sub { "https://api.forecast.io/forecast" },
);

has 'api_key' => (
    is => 'ro',
    required => 1
);


has 'ua' => (
    is => 'ro',
    default => sub {
        HTTP::Tiny->new(
            agent => "WebService::ForecastIO/$WebService::ForecastIO::VERSION ",
            SSL_options => {
                SSL_hostname => "",
                SSL_verify_mode => 0
            },
        );
    },
    lazy => 1,
);


has 'decoder' => (
    is => 'ro',
    default => sub {
        JSON->new();
    },
    lazy => 1,
);

sub request {
    my $self = shift;

    my $url = $self->base_url . "/" . $self->api_key . "/" . (join ",", @_);

    my $qp = join "&", (map {; "$_=" . $self->$_() } 
                        grep {; defined $self->$_() } qw(exclude units));

    if ( $qp ) {
        $url .= "?$qp";
    }

    my $response = $self->ua->get($url);

    if ( $response->{success} ) {
        $self->decoder->decode($response->{content});
    }
    else {
        die "Request to $url returned $response->{status}: $response->{content}\n";
    }
}

1;

__END__
=pod

=head1 NAME

WebService::ForecastIO::Request - Request role for WebService::ForecaseIO

=head1 VERSION

version 0.01

=head1 OVERVIEW

This is a role which implements requests to the L<forecast.io> API.

=head1 ATTRIBUTES

=head2 base_url

The base url to connect to the web service. Defaults to L<https://api.forecast.io/forecast>

=head2 ua

The user agent for the role. Uses L<HTTP::Tiny>.

=head2 decoder

The library to deserialize JSON responses. Uses L<JSON>.

=head1 AUTHOR

Mark Allen <mrallen1@yahoo.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Mark Allen.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



lib/WebService/ForecastIO.pm
use 5.014;

package WebService::ForecastIO;
{
  $WebService::ForecastIO::VERSION = '0.01';
}

use Moo;
with 'WebService::ForecastIO::Request';

use Time::Piece;

# ABSTRACT: Perl client for api.forecast.io


has 'units' => (
    is => 'ro',
);


has 'exclude' => (
    is => 'ro',
);


sub to_timepiece {
    my ($self, $epoch_secs) = @_;

    return Time::Piece->new($epoch_secs);
}


sub parse_datetime {
    my ($self, $iso8601) = @_;

    return Time::Piece->strptime($iso8601, "%Y-%m-%dT%H:%M:%S");
}


1;

__END__
=pod

=head1 NAME

WebService::ForecastIO - Perl client for api.forecast.io

=head1 VERSION

version 0.01

=head1 SYNOPSIS

    use 5.014;

    my $forecast = WebService::ForecastIO->new(
            api_key => 'secret',
            units => 'si', # metric units
            exclude => 'hourly,minutely,currently,flags', # get 7 day forecast only
    );

    #                             latitude          longitude
    my $json = $forecast->request(29.7492738082192, -95.4709680410959);

    # Human readable summary of daily data
    say $json->{daily}->{summary};

=head1 OVERVIEW

This is a Perl client for L<forecast.io API|https://developer.forecast.io>. Forecast.io 
applies "big data" analysis techniques to publicly available weather data including 
radar image analysis.  One of the things it attempts to predict is I<when> certain
weather events like rain will start, and the duration of those events.

This library requires an API key which can be obtained for free from the 
L<developer web site|https://developer.forecast.io>. The first 1,000 calls
per day are allowed without charge.  (More calls can be made if payment arrangements
are made.)

See the L<API docs|https://developer.forecast.io/docs/v2> for full details about
what data is provided and what granularity data sets are offered.

B<NOTE>: Errors are fatal. Please use something like L<Try::Tiny> if you
want to handle errors some other way.

=head1 ATTRIBUTES

=head2 units

This attribute overrides the default units of C<us>. Available units are:

=over

=item * C<si>

Metric units (windspeed in meters/second)

=item * C<ca>

Metric units, except windspeed in km/h

=item * C<uk>

Metric units, except windspeed in mi/h

=item * C<auto>

Unit selection based on request IP geolocation

=back

=head2 exclude

A comma delimited list of data sets to exclude. Choices are:

C<minutely>, C<currently>, C<hourly>, C<daily>, C<alerts>, C<flags>

See the API docs for information about what these data blocks represent.

Spaces should be omitted between exclusions.

=head2 api_key

Holds the API key for the forecast.io service.

=head1 METHODS

=head2 request

This method takes two or three parameters.

The first two parameters are C<latitude> and C<longitude> expressed
as floats. The optional third parameter is a C<time> value. The time
value can be expressed as either epoch seconds or as an iso8601
format time. 

Data sets up to 60 years old are available. (Such times should
be expressed as iso8601 format times since epoch seconds start
at midnight, January 1, 1970.)

=head2 to_timepiece

A convenience method that takes epoch seconds and returns a
new L<Time::Piece> object.  You can stringify this object to
iso8601 format by calling the C<datetime()> method on it.

=head2 parse_datetime

A convenience method that takes an iso8601 formatted string
and returns a new L<Time::Piece> object. This object can output
epoch seconds by calling the C<epoch()> method on it.

=head1 SEE ALSO

=over

=item * L<WebService::ForecastIO::Request>

=item * L<API docs|https://developer.forecast.io/docs/v2>

=back

=head1 AUTHOR

Mark Allen <mrallen1@yahoo.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Mark Allen.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut

